### PR TITLE
feat(nrecho-v4/v3): add New Relic instrumentation for Echo v4/v3

### PR DESCRIPTION
**Pull request by mirackara**
Add Support for nrecho v3 & v4

### CODE REVIEW

File: integrations/nrecho-v3/example/go.mod [hunk 1]
```diff
@@ -0,0 +1,22 @@
+module github.com/labstack/echo/examples
+
+go 1.25
+
+require github.com/labstack/echo/v4 v4.13.3
```
Comment (cade-conklin) [inline integrations/nrecho-v3/example/go.mod]: Is this supposed to be v3?  Would that allow you to get rid of the go.mod in the basic folder?

File: integrations/nrecho-v3/echo.go [hunk 2]
```diff
@@ -0,0 +1,210 @@
...
+// InstrumentEchoFunction verifies echo function calls and initiates tracing.
+// If tracing was added, then DefineTxnFromEchoCtx is called to inject the transaction
+// into the function body via the echo context.
+func InstrumentEchoFunction(manager *parser.InstrumentationManager, c *dstutil.Cursor) {
+	currentNode := c.Node()
+	switch v := currentNode.(type) {
+	case *dst.FuncDecl:
+		ctxName := GetEchoContextFromHandler(v.Type, manager.GetDecoratorPackage())
+		if ctxName == "" {
+			return
```
Comment (cade-conklin) [inline integrations/nrecho-v3/echo.go:215]: Question: Do we silently fail or should we returning an error for any reason?

File: integrations/nrecho-v3/echo.go [hunk 3]
```diff
@@ -0,0 +1,210 @@
...
+	}
+
+	return ""
+}
+
+// GetEchoContextFromHandler checks the type of a function or function literal declaration to determine if
+// this is an Echo handler. Returns the context variable of the echo handler.
+func GetEchoContextFromHandler(nodeType *dst.FuncType, pkg *decorator.Package) string {
+	// echo functions should only have 1 parameter
+	if len(nodeType.Params.List) != 1 {
```
Comment (cade-conklin) [inline integrations/nrecho-v3/echo.go:49]: Question: Do echo functions always have one parameter or can they have 0?

File: integrations/nrecho-v3/parsing_test.go [hunk 4]
```diff
@@ -0,0 +1,399 @@
...
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			defer parser.PanicRecovery(t)
+			got := nrecho.EchoMiddlewareCall(tt.stmt)
+			assert.Equal(t, tt.want, got)
+		})
+	}
+}
+
+func TestGetEchoHandlerContext(t *testing.T) {
```
Comment (cade-conklin) [inline integrations/nrecho-v3/parsing_test.go:256]: Chore: Add test cases for empty echo handler, and echo handler with 1 parameter, but varying lengths of names (0, 2 etc.)

File: integrations/nrecho-v3/parsing_test.go
Comment (cade-conklin) [inline integrations/nrecho-v3/parsing_test.go:1]: chore: Add a test for `hasExistingEchoTransaction`.  Needs cases for funcDecl being nil, funcDecl.Body being nil and different funcDecl shapes.  Let me know if it is too much of a lift though with how complicated the node object may turn out

File: integrations/nrecho-v3/echo.go [hunk 5]
```diff
@@ -0,0 +1,210 @@
...
+				if callExpr, ok := rhs.(*dst.CallExpr); ok {
+					if ident, ok := callExpr.Fun.(*dst.Ident); ok {
+						if ident.Name == "FromContext" && ident.Path == NrechoImportPath {
+							hasTransaction = true
+							return false
+						}
+					}
+				}
+			}
+		}
```
Comment (cade-conklin) [inline integrations/nrecho-v3/echo.go]: suggestion: Turn this into a function to make more testable.  Also un-nest the triple if statement to be more idiomatic go.

```suggestion
		func isFromContext() bool {...}
		
		dstutil.Apply(funcDecl.Body, func(c *dstutil.Cursor) bool {
                if isFromContext(c.Node()) {
                    hasTransaction = true
                    return false
               }
               return true
        }, nil)

```

File: integrations/nrecho-v3/parsing_test.go [hunk 6]
```diff
@@ -0,0 +1,399 @@
...
+		t.Run(tt.name, func(t *testing.T) {
+			ok := nrecho.GetEchoContextFromHandler(tt.node, tt.pkg)
+			if ok != tt.want {
+				t.Errorf("expected %v, got %v", tt.want, ok)
+			}
+		})
+	}
+}
+
+func TestDefineTxnFromEchoCtx(t *testing.T) {
```
Comment (cade-conklin) [inline integrations/nrecho-v3/parsing_test.go:436]: chore: Add more test cases, like an empty `body.List`and different length.

**Review by cade-conklin — CHANGES_REQUESTED** [review]
Only looked at a few of the files for v3 but I think some of these changes apply to v4 as well.  I will take a look at the rest tommorrow!

File: integrations/nrecho-v3/echo.go [hunk 7]
```diff
@@ -0,0 +1,210 @@
...
+	}
+	if call, ok := v.Rhs[0].(*dst.CallExpr); ok {
+		if ident, ok := call.Fun.(*dst.Ident); ok {
+			if ident.Name == "New" && ident.Path == echoImportPath {
+				if v.Lhs != nil {
+					return v.Lhs[0].(*dst.Ident).Name
+				}
+			}
+		}
+	}
```
Comment (cade-conklin) [inline integrations/nrecho-v3/echo.go:31]: suggestion: un-nest if statements and make into multiple `return ""`.

File: integrations/nrecho-v3/echo.go
Comment (cade-conklin) [inline integrations/nrecho-v3/echo.go:1]: suggestion: Add in some more comments about what different variables mean and what if statements might be checking

File: integrations/nrecho-v3/echo.go [hunk 8]
```diff
@@ -0,0 +1,210 @@
...
+
+// hasExistingEchoMiddleware checks if nrecho.Middleware is already present after the current router assignment
+func hasExistingEchoMiddleware(c *dstutil.Cursor) bool {
+	// Check all remaining statements to see if middleware is already added
+	parent := c.Parent()
+	if blockStmt, ok := parent.(*dst.BlockStmt); ok {
+		// Find current statement index
+		currentIndex := -1
+		for i, stmt := range blockStmt.List {
+			if stmt == c.Node() {
```
Comment (cade-conklin) [inline integrations/nrecho-v3/echo.go]: question: Are we checking if the contents of `stmt` and `c.Node()` are the same or if they are the same memory address?

File: integrations/nrecho-v3/echo.go [hunk 9]
```diff
@@ -0,0 +1,210 @@
...
+										}
+									}
+								}
+							}
+						}
+					}
+				}
+			}
+		}
+	}
```
Comment (cade-conklin) [inline integrations/nrecho-v3/echo.go]: suggestion: Break into its own testable function.  Add comments for each if statement and un-nest if statements for more idiomatic go

File: integrations/nrecho-v3/parsing_test.go
Comment (cade-conklin) [inline integrations/nrecho-v3/parsing_test.go:1]: chore: Add a test for `hasExistingEchoMiddleware` and the broken out function within

File: integrations/nrecho-v3/echo.go [hunk 10]
```diff
@@ -0,0 +1,210 @@
...
+	parent := c.Parent()
+	if blockStmt, ok := parent.(*dst.BlockStmt); ok {
+		// Find current statement index
+		currentIndex := -1
+		for i, stmt := range blockStmt.List {
+			if stmt == c.Node() {
+				currentIndex = i
+				break
+			}
+		}
```
Comment (cade-conklin) [inline integrations/nrecho-v3/echo.go]: fix: c.Index() already keeps track of the index so you don't need to find it

**Review by cade-conklin — CHANGES_REQUESTED** [review]

File: integrations/nrecho-v3/echo.go (hunk 2 above)
Comment (cade-conklin) [reply integrations/nrecho-v3/echo.go:215]: Nvm I see the difference between stateless and stateful functions

File: integrations/nrecho-v3/echo.go (hunk 3 above)
Comment (mirackara) [reply integrations/nrecho-v3/echo.go:49]: They will always have one parameter (the echo context). in order to register a echo route it must match the pattern as well

**Review by cade-conklin — APPROVED** [review]
LGTM!
